File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.swift
func solution() {
    guard let inputOfNMR = readLine() else { return }
    let arrayOfNMR: [Int] = inputOfNMR.split(separator: " ").compactMap { Int($0) }
                                                                        
    // N: number of nodes
    // M: number of edges
    // R: start node
    let (N, M, R) = (arrayOfNMR[0], arrayOfNMR[1], arrayOfNMR[2])

    // edges[i]: 2-Dimensional array of adjacent nodes of node i
    var edges = [[Int]](
        repeating: [Int](),
        count: N+1
    )
                                                                         
    for _ in 0..<M {
        guard let inputOfUV = readLine() else { return }
        let arrayOfUV: [Int] = inputOfUV.split(separator: " ").compactMap { Int($0) }
        let (u, v) = (arrayOfUV[0], arrayOfUV[1])
        edges[u].append(v)
        edges[v].append(u)
    }
    
    // Start DFS
    print(dfs(R, &edges).map { String($0) }.joined(separator: "\n"))
}


func dfs(
    _ v: Int,
    _ edges: inout [[Int]]
) -> [Int] {
    var visited = [Int](
        repeating: 0,
        count: edges.count
    )
    var order: Int = 1
    var stack: [Int] = [v]

    while !stack.isEmpty {
        let edge: Int = stack.removeLast()

        if visited[edge] == 0 {
            visited[edge] = order
            order += 1

            for x in edges[edge].sorted() {
                if visited[x] == 0 {
                    stack.append(x)
                }
            }
        }
    }

    visited.removeFirst()
    
    return visited
}

solution()

File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.py
import sys

input = sys.stdin.readline

# N: number of nodes
# M: number of edges
# R: start node
N, M, R = map(int, input().split())

# edges[i]: list of adjacent nodes of node i
edges = [[] for _ in range(N + 1)]
for _ in range(M):
    u, v = map(int, input().split())
    edges[u].append(v)
    edges[v].append(u)

def dfs(v):
    # visited[i]: order when node i was visited (0 if not visited)
    visited = [0 for _ in range(N + 1)]
    stack = [v]
    order = 1

    while stack:
        node = stack.pop()
        if visited[node] == 0:
            visited[node] = order
            order += 1

           # Push adjacent nodes in ascending order
            # so that higher-numbered nodes are visited first (due to stack LIFO)
            for x in sorted(edges[node]):
                if visited[x] == 0:
                    stack.append(x)
                
    return visited

# Run DFS and print visit order for each node (1-indexed)
print("\n".join(map(str, dfs(R)[1:])))